File: src/infrastructure/spreadsheet/purchaseSheet.js
/* exported PurchaseSheet */

class PurchaseSheet extends BaseSheet {
  constructor(sheetName){
    // 仕入管理シートは4行目がヘッダー前提
    super(sheetName, 4, 1);
    this.rowNumbers = [];
  }

  setRowNumbers(rowNumbers){
    this.rowNumbers = rowNumbers;
  }

  filter(columnName, values) {
    console.log(`Filtering column "${columnName}" with values: ${JSON.stringify(values)}`);
    const columnIndex = this._getColumnIndexByName(columnName);
    const rowNumbers = [];
    const filteredData = [];
    
    for (let i = 0; i < this.data.length; i++) {
      const rowValue = String(this.data[i][columnIndex]);
      if (values.some(v => String(v) === rowValue)) {
        // BaseRow.rowNumber は実シート上の行番号
        rowNumbers.push(this.data[i].rowNumber);
        filteredData.push(this.data[i]);
      }
    }
    
    this.rowNumbers = rowNumbers;
    this.data = filteredData;
    console.log(`${columnName}でフィルタリング: ${rowNumbers.length}行が見つかりました`);
    return this.data;
  }

  writeColumn(columnName, value){
    try {
      const column = this._getColumnIndexByName(columnName) + 1;
      let successCount = 0;
      
      for (const rowNum of this.rowNumbers) {
          if (typeof value === 'object' && value.type === 'formula') {
            this.sheet.getRange(rowNum, column).setFormula(value.value);
          } else {
            this.sheet.getRange(rowNum, column).setValue(value);
          }
          successCount++;
      }
      console.log(`${columnName}を${successCount}行に書き込みました`);
    } catch (e) {
      console.warn(`${columnName}への書き込みに失敗しました: ${e.message}`);
      throw e;
    }
  }

  _generatePlanNameText() {
    const dateStr = this._formatDateMMDD(new Date());
    let deliveryCategory = '';
    try {
      deliveryCategory = this.data.length > 0 ? (this.data[0].get("納品分類") || '') : '';
    } catch (e) {
      deliveryCategory = '';
    }
    return `${dateStr}${deliveryCategory}`;
  }

  _formatDateMMDD(date) {
    const month = String(date.getMonth() + 1);
    const day = String(date.getDate());
    const monthStr = month.length === 1 ? '0' + month : month;
    const dayStr = day.length === 1 ? '0' + day : day;
    return `${monthStr}/${dayStr}`;
  }

  aggregateItems() {
    const aggregatedItems = {};
    const labelOwner = 'SELLER';
    
    for (let i = 0; i < this.data.length; i++) {
      const row = this.data[i];
      let sku = '';
      let asin = '';
      let quantity = 0;
      try { sku = row.get("SKU") || ''; } catch (e) {}
      try { asin = row.get("ASIN") || ''; } catch (e) {}
      try { quantity = Number(row.get("購入数")); } catch (e) { quantity = 0; }
      
      if (!sku || !quantity || quantity <= 0) {
        console.warn(`納品プラン対象外: sku=${sku}, quantity=${quantity}`);
        continue;
      }
      
      if (!aggregatedItems[sku]) {
        aggregatedItems[sku] = {
          msku: sku,
          asin: asin,
          quantity: 0,
          labelOwner: labelOwner
        };
      }
      aggregatedItems[sku].quantity += quantity;
    }
    
    return aggregatedItems;
  }

  writePlanResult(planResult) {
    const planCol = this._getColumnIndexByName("納品プラン") + 1;
    const shipDateCol = this._getColumnIndexByName("発送日") + 1;

    const link = planResult && planResult.link ? String(planResult.link) : '';
    const inboundPlanId = planResult && planResult.inboundPlanId ? String(planResult.inboundPlanId) : '';

    const dateOnly = new Date();
    dateOnly.setHours(0, 0, 0, 0);
    const dateStr = Utilities.formatDate(dateOnly, 'Asia/Tokyo', 'yyyy/MM/dd');

    for (const rowNum of this.rowNumbers) {
      // 納品プラン
      if (link) {
        const displayText = inboundPlanId || this._generatePlanNameText();
        const linkFormula = `=HYPERLINK("${link}", "${displayText}")`;
        const r = this.sheet.getRange(rowNum, planCol);
        r.setFormula(linkFormula);
        console.log(`[納品プラン] wrote ${r.getA1Notation()} formula=${linkFormula}`);
      } else if (inboundPlanId) {
        const r = this.sheet.getRange(rowNum, planCol);
        r.setValue(inboundPlanId);
        console.log(`[納品プラン] wrote ${r.getA1Notation()} value=${inboundPlanId}`);
      } else {
        console.log(`[納品プラン] skip row=${rowNum} (no link/inboundPlanId)`);
      }

      // 発送日
      const d = this.sheet.getRange(rowNum, shipDateCol);
      d.setValue(dateOnly);
      console.log(`[発送日] wrote ${d.getA1Notation()} value=${dateStr}`);
    }
  }

  decreasePurchaseQuantity(quantity) {
    const quantityColumnIndex = this._getColumnIndexByName('購入数');
    let successCount = 0;
    const zeroQuantityRows = [];

    for (const rowNum of this.rowNumbers) {
      const currentQuantity = Number(this.sheet.getRange(rowNum, quantityColumnIndex + 1).getValue());
      const newQuantity = Math.max(0, currentQuantity - quantity);
      this.sheet.getRange(rowNum, quantityColumnIndex + 1).setValue(newQuantity);
      successCount++;
      console.log(`行${rowNum}: 購入数を${currentQuantity}から${newQuantity}に減らしました`);

      if (newQuantity === 0) {
        zeroQuantityRows.push(rowNum);
      }
    }

    console.log(`${successCount}行の購入数を${quantity}減らしました`);
    return zeroQuantityRows;
  }

  deleteRows(rowNumbers) {
    if (!rowNumbers || rowNumbers.length === 0) {
      return;
    }

    const sortedRows = rowNumbers.slice().sort((a, b) => b - a);

    for (const rowNum of sortedRows) {
      this.sheet.deleteRow(rowNum);
      console.log(`行${rowNum}を削除しました`);
    }

    console.log(`${sortedRows.length}行を削除しました`);
  }

  /**
   * FNSKU補完（SKU -> FNSKU）
   * @param {string} accessToken
   * @param {BaseRow[]|null} targetRows 省略時はthis.data全体。指定時はその行だけ補完する。
   */
  fetchMissingFnskus(accessToken, targetRows = null) {
    const fnskuGetter = new FnskuGetter(accessToken);
    const fnskuColIndex = this._getColumnIndexByName("FNSKU");
    const fnskuCol = fnskuColIndex + 1;

    const strict = Array.isArray(targetRows);
    const rowsToProcess = strict ? targetRows : this.data;

    for (const row of rowsToProcess) {
      const fnsku = row.get("FNSKU");
      const sku = row.get("SKU");
      const rowNum = row.rowNumber;

      if (!fnsku || fnsku === '') {
        // #region agent log
        try{UrlFetchApp.fetch('http://127.0.0.1:7243/ingest/24045cd5-2584-4703-ab97-0442e02ed8a6',{method:'post',contentType:'application/json',payload:JSON.stringify({location:'purchaseSheet.js:PurchaseSheet.fetchMissingFnskus:before',message:'about to call getFnsku',data:{rowNum:rowNum,skuRaw:String(sku||''),skuTrim:String(sku||'').trim(),skuHasWhitespace:/\\s/.test(String(sku||'')),skuHasLeadingOrTrailingWhitespace:/^\\s|\\s$/.test(String(sku||''))},timestamp:Date.now(),sessionId:'debug-session',runId:'pre-fix',hypothesisId:'B'})});}catch(e){}
        try {
          const raw = String(sku ?? '');
          console.log(`[DBG][FNSKU][fetchMissingFnskus] rowNum=${rowNum} skuRaw=${JSON.stringify(raw)} skuTrim=${JSON.stringify(raw.trim())} skuCharCodesHead=${JSON.stringify(raw.slice(0,50).split('').map(c=>c.charCodeAt(0)))}`);
        } catch (e) {}
        // #endregion

        const skuTrim = String(sku || '').trim();
        if (!skuTrim) {
          console.warn(`[FNSKU補完] SKUが空のためFNSKU補完できません: row=${rowNum}`);
          if (strict) {
            throw new Error(`SKUが空のためFNSKU補完できません。行${rowNum}のSKU（またはASIN->SKU補完）を確認してください。`);
          }
          continue;
        }

        const fetchedFnsku = fnskuGetter.getFnsku(skuTrim);
        console.log(`Fetched FNSKU for ${sku}: ${fetchedFnsku}`);
        if (rowNum && fnskuCol >= 1) {
          this.writeCell(rowNum, fnskuCol, fetchedFnsku);
          // 指示書作成は this.data を参照するため、メモリ上の行データも更新する
          row[fnskuColIndex] = fetchedFnsku;
        }
      }
    }
  }

  /**
   * 指示書作成前の補完: SKUが空の場合、ASINから出品レポートでSKUを解決して書き戻す
   * @param {string} accessToken
   * @param {BaseRow[]|null} targetRows 省略時はthis.data全体が対象。指定時はその行だけを補完する（参照はシート全体）。
   */
  fillMissingSkusFromAsins(accessToken, targetRows = null) {
    const resolver = new MerchantListingsSkuResolver(accessToken);
    const skuColIndex = this._getColumnIndexByName("SKU");
    const skuCol = skuColIndex + 1;

    const asinToSkuLocal = this._buildAsinToSkuLocalMap_();
    const { targets } = this._collectMissingSkuTargets_(targetRows);

    if (targets.length === 0) {
      console.log('[SKU補完] SKU空白なし -> スキップ');
      return;
    }

    // 1) 仕入管理シート内の既存行のSKUを流用
    let filledByLocal = 0;
    const stillMissingAsins = [];
    for (const t of targets) {
      const resolvedLocal = asinToSkuLocal.get(t.asin);
      if (!resolvedLocal) {
        stillMissingAsins.push(t.asin);
        continue;
      }
      if (t.rowNum && skuCol >= 1) {
        this.writeCell(t.rowNum, skuCol, resolvedLocal);
        t.row[skuColIndex] = resolvedLocal; // in-memoryも更新
        filledByLocal++;
      }
    }

    // 2) まだ無い分だけ Reports API で解決
    const uniqueMissing = Array.from(new Set(stillMissingAsins));
    if (uniqueMissing.length === 0) {
      console.log(`[SKU補完] ローカル流用で全件補完: ${filledByLocal}/${targets.length}`);
      return;
    }

    console.log(`[SKU補完] ローカル流用=${filledByLocal}/${targets.length} -> Reports APIで解決開始 (ASIN数=${uniqueMissing.length})`);
    const asinToSkuRemote = resolver.resolveSkusByAsins(uniqueMissing);

    let filledByRemote = 0;
    for (const t of targets) {
      const currentSku = String(t.row.get("SKU") || '').trim();
      if (currentSku) continue;
      const resolved = asinToSkuRemote.get(t.asin);
      if (!resolved) continue;
      if (t.rowNum && skuCol >= 1) {
        this.writeCell(t.rowNum, skuCol, resolved);
        t.row[skuColIndex] = resolved;
        filledByRemote++;
      }
    }
    console.log(`[SKU補完] 書き込み完了: local=${filledByLocal}, remote=${filledByRemote}, total=${filledByLocal + filledByRemote}/${targets.length}`);
  }

  /**
   * 仕入管理シート全体から「ASIN -> SKU」のローカル辞書を作る（同一ASINがあれば既存SKUを流用）
   * @returns {Map<string,string>}
   */
  _buildAsinToSkuLocalMap_() {
    const asinToSkuLocal = new Map();
    const sourceRows = Array.isArray(this.allData) ? this.allData : this.data;
    for (const row of sourceRows) {
      const sku = String(row.get("SKU") || '').trim();
      const asin = String(row.get("ASIN") || '').trim();
      if (asin && sku && !asinToSkuLocal.has(asin)) {
        asinToSkuLocal.set(asin, sku);
      }
    }
    return asinToSkuLocal;
  }

  /**
   * SKUが空の行だけを対象に集める（ASIN有無は問わない）
   * @param {BaseRow[]|null} targetRows
   * @returns {{targets: Array<{row: any, rowNum: number, asin: string}>}}
   */
  _collectMissingSkuTargets_(targetRows) {
    const targets = []; // { row, rowNum, asin }
    const rowsToProcess = Array.isArray(targetRows) ? targetRows : this.data;
    for (const row of rowsToProcess) {
      const rowNum = row.rowNumber;
      const sku = String(row.get("SKU") || '').trim();
      const asin = String(row.get("ASIN") || '').trim();
      if (!sku) {
        targets.push({ row, rowNum, asin });
      }
    }
    return { targets };
  }

  writePlanNameToRows(instructionURL) {
    const dateStr = this._formatDateMMDD(new Date());

    return this.writeColumnByFunc("プラン別名", (row) => {
      const deliveryCategory = row.get("納品分類") || '';
      const planNameValue = `${dateStr}${deliveryCategory}`;

      if (instructionURL) {
        return { type: 'formula', value: `=HYPERLINK("${instructionURL}", "${planNameValue}")` };
      }
      return planNameValue;
    });
  }

}

File: python/src/infrastructure/spreadsheet/purchase_sheet.py
from __future__ import annotations

import logging
from datetime import datetime
from typing import Any

from infrastructure.amazon.fnsku_getter import FnskuGetter
from infrastructure.amazon.merchant_listings_sku_resolver import MerchantListingsSkuResolver
from infrastructure.spreadsheet.base_sheet import BaseSheet
from infrastructure.spreadsheet.base_sheets_repository import BaseSheetsRepository

logger = logging.getLogger(__name__)

HEADER_ROW = 4


class PurchaseSheet(BaseSheet):
    def __init__(self, repo: BaseSheetsRepository, sheet_id: str, sheet_name: str) -> None:
        super().__init__(repo=repo, sheet_id=sheet_id, sheet_name=sheet_name, header_row=HEADER_ROW)

    def aggregate_items(self) -> dict[str, dict[str, Any]]:
        aggregated: dict[str, dict[str, Any]] = {}
        label_owner = "SELLER"
        for row in self.data:
            try:
                sku = str(row.get("SKU") or "").strip()
            except Exception:
                sku = ""
            try:
                asin = str(row.get("ASIN") or "").strip()
            except Exception:
                asin = ""
            try:
                quantity = int(row.get("購入数") or 0)
            except Exception:
                quantity = 0
            if not sku or quantity <= 0:
                logger.warning("納品プラン対象外: sku=%s, quantity=%d", sku, quantity)
                continue
            if sku not in aggregated:
                aggregated[sku] = {"msku": sku, "asin": asin, "quantity": 0, "labelOwner": label_owner}
            aggregated[sku]["quantity"] += quantity
        return aggregated

    def write_plan_result(self, plan_result: dict[str, Any]) -> None:
        plan_col = self._get_column_index_by_name("納品プラン") + 1
        ship_date_col = self._get_column_index_by_name("発送日") + 1
        link = str(plan_result.get("link", ""))
        inbound_plan_id = str(plan_result.get("inboundPlanId", ""))
        today = datetime.now().strftime("%Y/%m/%d")
        for row in self.data:
            row_num = row.row_number
            if link:
                display = inbound_plan_id or self._generate_plan_name_text()
                formula = f'=HYPERLINK("{link}", "{display}")'
                self.write_formula(row_num, plan_col, formula)
            elif inbound_plan_id:
                self.write_cell(row_num, plan_col, inbound_plan_id)
            self.write_cell(row_num, ship_date_col, today)

    def decrease_purchase_quantity(self, quantity: int) -> list[int]:
        qty_col = self._get_column_index_by_name("購入数") + 1
        zero_rows: list[int] = []
        for row in self.data:
            current = int(self._worksheet.cell(row.row_number, qty_col).value or 0)
            new_qty = max(0, current - quantity)
            self.write_cell(row.row_number, qty_col, new_qty)
            logger.info("行%d: 購入数を%dから%dに減らしました", row.row_number, current, new_qty)
            if new_qty == 0:
                zero_rows.append(row.row_number)
        return zero_rows

    def delete_rows(self, row_numbers: list[int]) -> None:
        for row_num in sorted(row_numbers, reverse=True):
            self._worksheet.delete_rows(row_num)
            logger.info("行%dを削除しました", row_num)

    def fetch_missing_fnskus(self, access_token: str) -> None:
        getter = FnskuGetter(access_token)
        fnsku_col_idx = self._get_column_index_by_name("FNSKU")
        fnsku_col = fnsku_col_idx + 1
        for row in self.data:
            fnsku = str(row.get("FNSKU") or "").strip()
            sku = str(row.get("SKU") or "").strip()
            if not fnsku and sku:
                fetched = getter.get_fnsku(sku)
                logger.info("FNSKU取得: %s -> %s", sku, fetched)
                self.write_cell(row.row_number, fnsku_col, fetched)
                row[fnsku_col_idx] = fetched

    def fill_missing_skus_from_asins(self, access_token: str) -> None:
        sku_col_idx = self._get_column_index_by_name("SKU")
        sku_col = sku_col_idx + 1
        asin_to_sku_local = self._build_asin_to_sku_local_map()
        targets = self._collect_missing_sku_targets()
        if not targets:
            logger.info("[SKU補完] SKU空白なし -> スキップ")
            return
        filled_local = 0
        still_missing_asins: list[str] = []
        for t in targets:
            resolved = asin_to_sku_local.get(t["asin"])
            if resolved:
                self.write_cell(t["row_num"], sku_col, resolved)
                t["row"][sku_col_idx] = resolved
                filled_local += 1
            else:
                still_missing_asins.append(t["asin"])
        unique_missing = list(set(still_missing_asins))
        if not unique_missing:
            logger.info("[SKU補完] ローカル流用で全件補完: %d/%d", filled_local, len(targets))
            return
        logger.info("[SKU補完] ローカル=%d/%d -> Reports API (ASIN数=%d)", filled_local, len(targets), len(unique_missing))
        resolver = MerchantListingsSkuResolver(access_token)
        asin_to_sku_remote = resolver.resolve_skus_by_asins(unique_missing)
        filled_remote = 0
        for t in targets:
            current = str(t["row"].get("SKU") or "").strip()
            if current:
                continue
            resolved = asin_to_sku_remote.get(t["asin"])
            if resolved:
                self.write_cell(t["row_num"], sku_col, resolved)
                t["row"][sku_col_idx] = resolved
                filled_remote += 1
        logger.info("[SKU補完] 完了: local=%d, remote=%d, total=%d/%d", filled_local, filled_remote, filled_local + filled_remote, len(targets))

    def write_plan_name_to_rows(self, instruction_url: str | None) -> int:
        date_str = self._format_date_mmdd()
        def value_func(row: Any, _index: int) -> Any:
            delivery_category = str(row.get("納品分類") or "").strip()
            plan_name = f"{date_str}{delivery_category}"
            if instruction_url:
                return {"type": "formula", "value": f'=HYPERLINK("{instruction_url}", "{plan_name}")'}
            return plan_name
        return self.write_column_by_func("プラン別名", value_func)

    def _build_asin_to_sku_local_map(self) -> dict[str, str]:
        result: dict[str, str] = {}
        for row in self.all_data:
            sku = str(row.get("SKU") or "").strip()
            asin = str(row.get("ASIN") or "").strip()
            if asin and sku and asin not in result:
                result[asin] = sku
        return result

    def _collect_missing_sku_targets(self) -> list[dict[str, Any]]:
        targets: list[dict[str, Any]] = []
        for row in self.data:
            sku = str(row.get("SKU") or "").strip()
            asin = str(row.get("ASIN") or "").strip()
            if not sku:
                targets.append({"row": row, "row_num": row.row_number, "asin": asin})
        return targets

    def _generate_plan_name_text(self) -> str:
        date_str = self._format_date_mmdd()
        try:
            category = str(self.data[0].get("納品分類") or "").strip() if self.data else ""
        except Exception:
            category = ""
        return f"{date_str}{category}"

    def _format_date_mmdd(self) -> str:
        now = datetime.now()
        return f"{now.month:02d}/{now.day:02d}"
